rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -1030,39 +1036,121 @@ static rmw_ret_t discovery_thread_stop(rmw_dds_common::Context & common_context)
+static bool check_create_domain(dds_domainid_t did, rmw_discovery_options_t * discovery_options)
 {
-  const bool localhost_only = (localhost_only_option == RMW_LOCALHOST_ONLY_ENABLED);
   std::lock_guard<std::mutex> lock(gcdds().domains_lock);
-  /* return true: n_nodes incremented, localhost_only set correctly, domain exists
+  /* return true: n_nodes incremented, discovery params set correctly, domain exists
      "      false: n_nodes unchanged, domain left intact if it already existed */
   CddsDomain & dom = gcdds().domains[did];
   if (dom.refcount != 0) {
-    /* Localhost setting must match */
-    if (localhost_only == dom.localhost_only) {
+    /* Discovery parameters must match */
+    bool options_equal = false;
+    const auto rc = rmw_discovery_options_equal(discovery_options, &dom.discovery_options, &options_equal);
+    if (RMW_RET_OK != rc) {
+      RCUTILS_LOG_ERROR_NAMED(
+        "rmw_cyclonedds_cpp",
+        "check_create_domain: unable to check if discovery options are equal: %i",
+        rc);
+      return false;
+    }
+    if (options_equal) {
       dom.refcount++;
       return true;
     } else {
       RCUTILS_LOG_ERROR_NAMED(
         "rmw_cyclonedds_cpp",
-        "rmw_create_node: attempt at creating localhost-only and non-localhost-only nodes "
-        "in the same domain");
+        "check_create_domain: attempt at creating nodes in the same domain with different "
+        "discovery parameters");
       return false;
     }
   } else {
     dom.refcount = 1;
-    dom.localhost_only = localhost_only;
-
-    /* Localhost-only: set network interface address (shortened form of config would be
-       possible, too, but I think it is clearer to spell it out completely).  Empty
-       configuration fragments are ignored, so it is safe to unconditionally append a
-       comma. */
-    std::string config =
-      localhost_only ?
-      "<CycloneDDS><Domain><General><Interfaces><NetworkInterface address=\"127.0.0.1\"/>"
-      "</Interfaces></General></Domain></CycloneDDS>,"
-      :
-      "";
+    dom.discovery_options = *discovery_options;
+
+    std::string config = "<CycloneDDS><Domain>";
+    bool add_localhost_as_static_peer = false;
+    bool multicast_off = false;
+
+    switch (discovery_options->automatic_discovery_range) {
+      case RMW_AUTOMATIC_DISCOVERY_RANGE_SUBNET:
+        /* This is the default behavior of DDS and does not require a special
+           configuration. */
+        break;
+      default:
+        /* This situation would happen if a new option is introduced for the
+           ROS_AUTOMATIC_DISCOVERY_RANGE but rmw_cyclonedds_cpp is not updated
+           to handle it. */
+        /* TODO(MXG): Displaying an integer is not very informative for users.
+           It would be better if rmw knows how to translate the integer values
+           into their associated option strings. However the knowledge of the
+           string values is currently located in rcl instead of rmw. */
```

Comment:
Should we address this TODO?

Reply:
If I understand correctly, addressing this TODO would mean making an `enum -> string` conversion function at the RMW layer; however, if `rmw_cyclonedds_cpp` doesn't know the meaning of an enum value then either it's a new option, or it's an invalid value.

If it's a new option, then maybe the `rmw` library linked knows about it and can display something sensible (case 1). However, maybe the `rmw` library linked is also out of date and it doesn't know about the option so the string it displays is "unknown" (case 2). If it's an invalid value than the best a conversion function could do was say "unknown" (case 3). I think the integer value would be more valuable than "unknown" in cases 2 and 3, an inconvenient but the same amount of information in case 1.

I'd vote to remove the `TODO` instead of addressing it.

Reply:
Removed TODO in 5ed2f3b